doc/arch/adr-010-learn-to-rank.md

```diff
@@ -0,0 +1,286 @@
+  the search team has already worked with
+- TensorFlow provides an off-the-shelf ranking module, which solves
+  some of the machine learning problems of LTR for us
+- Manual relevancy tuning is not scalable, and won't get us to the same
+  level of relevancy as LTR
+
+## Implementation plan
+
+We have decided to implement Learning To Rank by following on from
+the work started in the [validation spike][].
+
+As we are at the beginning of the project, this section may not
+accurately reflect the final implementation.
+
+[validation spike]: https://github.com/alphagov/search-api/pull/1768
+
+### Language
+
+We would like to keep the code for this project in Ruby as much as
+possible. Search API is written in Ruby and it is the primary language
+used at GOV.UK. Necessarily some machine learning code will need to be
+written in Python.
+
+### Machine learning library
+
+We have decided to [TensorFlow's ranking library][] for our
+implementation of Learning To Rank.
+
+TensorFlow is well supported, and the ranking library is under active development.
+
+We considered using the [elasticsearch-learning-to-rank][] plugin.
+However, AWS Elasticsearch does not permit us to install custom plugins.
+
+We also considered PyTorch. However, this would require writing much more
+custom machine learning code, as PyTorch don't currently have a ranking library.
+
+[TensorFlow's ranking library]: https://github.com/tensorflow/ranking
+[elasticsearch-learning-to-rank]: https://github.com/o19s/elasticsearch-learning-to-rank
+
+### Dockerised model
+
+We will use TensorFlow Serving to make a re-ranker model available
+to Search API on search machines.
+
+We will run a docker container (containing our pre-trained model) on
+search machines to which Search API can send requests.
+
+Should the docker container fail, Search API should be resilient to this,
+and fall back to the default ranking algorithm, with results not
+re-ranked by the model.
+
+### Training
+
+**Training the model** should happen on search machines on a schedule.
+An instance of Search API will train a model using the latest relevancy
+judgements stored in S3, and will then upload a new version of the model to S3.
+On a schedule all search machines will pull the latest version of the model
+and push the new version to the locally running TF Serving container.
+
+**Training data** is made up of relevancy judgements and document features.
+The judgements may be sourced from user behaviour (implicit signals
+of relevancy) perhaps with a Click Model; so far we have used explicit
+manual judgements.
+
+Acquiring relevancy judgements is a work in progress. Our next project
+will be to create a process by which we can acquire relevancy judgements
+on a rolling basis automatically.
+
+### Re-ranking requests
+
+**Using the model** to re-rank results will involve a user searching,
+Search API will perform the query for the top-k results (say 100),
```

Comment:
We'll have to consider performance here and do appropriate testing so get confidence that we'll be able to support the extra load. I wonder if there's a way of measuring what the lowest ranked results that appear are, and adjust the window appropriately

Reply:
That's interesting. Perhaps we could do some analyses of the es_score@k to work out what the top-k size should be, or figure out a way to do it at query time. I added a task to the LTR epic for this.

There's a section on improving reranking performance in the second half of the Bloomberg talk on LTR: https://youtu.be/eMuepJpjUjI?t=1235.